### PR TITLE
feat: Support weight streaming

### PR DESCRIPTION
**Pull request by keehyuna**
# Description

- Weight streaming feature is exposed as compiler option to set percent or weight streaming bytes
- Create a network with kSTRONGLY_TYPED and set kWEIGHT_STREAMING to builder config
- Same dtypes are required for layers in strongly typed network

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 1]
```diff
@@ -109,10 +111,119 @@ def __init__(
         if self.serialized_engine is not None and not self.settings.lazy_engine_init:
             self.setup_engine()
 
+    def set_weight_streaming_budget(self) -> None:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:114]: @keehyuna do you need to add something similar to the C++ API?

File: core/runtime/TRTEngine.cpp [hunk 2]
```diff
@@ -78,8 +78,7 @@ TRTEngine::TRTEngine(
   cuda_engine = make_trt(rt->deserializeCudaEngine(serialized_engine.c_str(), serialized_engine.size()));
   TORCHTRT_CHECK((cuda_engine.get() != nullptr), "Unable to deserialize the TensorRT engine");
 
-  exec_ctx = make_trt(cuda_engine->createExecutionContext());
-  TORCHTRT_CHECK((exec_ctx.get() != nullptr), "Unable to create TensorRT execution context");
+  exec_ctx = nullptr;
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp:81]: Why did you remove the smart pointer here?

File: core/runtime/execute_engine.cpp [hunk 3]
```diff
@@ -95,11 +95,13 @@ bool _cudagraphs_validate_shapes(std::vector<at::Tensor> inputs, c10::intrusive_
 }
 
 std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intrusive_ptr<TRTEngine> compiled_engine) {
+  compiled_engine->init_context();
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:98]: This will add first run latency. Why cant it run in the constructor?

File: core/runtime/TRTEngine.cpp [hunk 4]
```diff
@@ -248,6 +245,32 @@ void TRTEngine::set_profiling_paths() {
   cuda_graph_debug_path = std::filesystem::path{profile_path_prefix + "/" + name + "_cudagraph.dot"}.string();
 }
 
+int64_t TRTEngine::get_weight_streaming_budget_v2() {
+  return cuda_engine->getWeightStreamingBudgetV2();
+}
+
+bool TRTEngine::set_weight_streaming_budget_v2(int64_t budget) {
+  return cuda_engine->setWeightStreamingBudgetV2(budget);
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp:253]: If a new weight stream budget requires a new execution context, that should be done atomically with this API

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -83,7 +83,7 @@ def compile(
     hardware_compatible: bool = _defaults.HARDWARE_COMPATIBLE,
     timing_cache_path: str = _defaults.TIMING_CACHE_PATH,
     lazy_engine_init: bool = _defaults.LAZY_ENGINE_INIT,
-    weight_streaming_setting: str = _defaults.STREAMABLE_WEIGHTS_DISABLED,
+    weight_streaming: bool = _defaults.WEIGHT_STREAMING,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:86]: `enable_weight_streaming`

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 6]
```diff
@@ -218,13 +219,25 @@ def set_weight_streaming_budget_v1(
                     self.engine.minimum_weight_streaming_budget
                 )
 
+    def reset_context(self):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:222]: Do these context resets atomically with whatever runtime settting change. Leave as much out of the forward function as we can

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 7]
```diff
@@ -104,16 +106,138 @@ def __init__(
...
+        if self.context != None:
+            del self.context
+            self.context = None
+
+    def lazy_init_context(self):
+        # Lazy creation of context when it is actually needed at forward() or enable_profiling()
+        assert self.engine, f"Context is used before setting up the engine"
+
+        if self.context is None:
+            self.context = self.engine.create_execution_context()
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:233]: we already have a setup engine function, not sure why we need to handle this at exec time?

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py (hunk 6 above)
Comment (keehyuna) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: I came crossed two ideas and tried #1. Please let me know if there is better way to handle it automatically.
1. reset_context(delete context) and  apply set_weight_streaming_budget() api. context is created at forward()
2. Enqueue runtime setting change like set_weight or profile enable. Then delete context->apply pending api-> create context in forward()

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 8]
```diff
@@ -104,16 +104,46 @@ def __init__(
...
+            budget_bytes = self.engine.weight_streaming_budget_v2
+        if self.engine.streamable_weights_size == budget_bytes:
+            logger.warning("Weight streaming is disabled")
+
+        return budget_bytes
+
+    def set_automatic_streaming_budget(self):
+        budget_bytes = self.engine.get_weight_streaming_automatic_budget()
+        return self.set_weight_streaming_budget(budget_bytes)
+
```
Comment (keehyuna) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:141]: This api is same as in [TorchTensorRTModule](https://github.com/pytorch/TensorRT/pull/3111/commits/2fe93697e9d7274968cc639ee51a9e360135dcac#diff-10ebe63ef48658946436e5c01d9b062c2527b2a8990a6b6ab8b067c1783571baR137-R154) class. If this interface is good to go, parent class can used to share it and other some methods.

**Comment by narendasan** [comment]
We probably need to think about what the user flow is here: 

So @ compile-time:
1. Users tell us to make an engine that is weight streamable

@ runtime
1. How do we set up the engine as a default? should there be a default weight budget? 
2. User now wants to explicitly set the engine weight budgets 
   1. How can they do this from the module level? What happens if there are multiple engines in the graph?
   2. We need to recreate the execution context, imo this needs to be done atomically with this call to keep it out of the forward function. i.e. as part of set_weight_budget we recreate the execution context. 
   3.

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 9]
```diff
@@ -191,6 +221,7 @@ def __del__(self) -> None:
             self.cudagraph.reset()
 
     def forward(self, *inputs: torch.Tensor) -> torch.Tensor | Tuple[torch.Tensor, ...]:
+        self.init_context()
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:224]: I really want to pull these calls out, It should assume that the engine is setup and error if not

Comment (keehyuna) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: Understood. recreation of context happens only when [set_weight_streaming_budget](https://github.com/pytorch/TensorRT/pull/3111/files#diff-10ebe63ef48658946436e5c01d9b062c2527b2a8990a6b6ab8b067c1783571baR162) is called

**Comment by keehyuna** [comment]
Hi @narendasan 

1. When only compiler option is provided,

>automatic weight streaming budget is applied. [test case](https://github.com/pytorch/TensorRT/pull/3111/files#diff-f9e9b91837ae150afd9f6637efb856ac105a78ec762851d0477c80a3e117dd82R44-R73)

2. Weight stream size is set manually
>torchtrt.runtime.weight_streaming_context() runtime api is added to get streamable size and set budget. [test case](https://github.com/pytorch/TensorRT/pull/3111/files#diff-f9e9b91837ae150afd9f6637efb856ac105a78ec762851d0477c80a3e117dd82R81-R119)
3. Context recreation
>added decorator in [set_weight_streaming_budget](https://github.com/pytorch/TensorRT/pull/3111/files#diff-0f9e88e05100eb9d4c2d0f180711a84d69305668735619a8754693ded1ea4901R145).
4. multiple subgraph
>I'm thinking of applying normalized size if multiple module is in runtime module. [impl](https://github.com/pytorch/TensorRT/pull/3111/files#diff-eb60cde007172b5238d6c8fe98f40b16816e500cd5b0cc08ac83cb5de1d77938R47-R57). Too small value will have bad impact. I will test various size.

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py [hunk 10]
```diff
@@ -129,6 +146,40 @@ def __init__(
...
+    def set_weight_streaming_budget(self, budget_bytes):
+        return self._set_weight_streaming_budget(budget_bytes)
+
+    def _set_weight_streaming_budget(self, budget_bytes):
+        # Disable weight streaming for invalid budget size
+        if budget_bytes <= 0:
+            budget_bytes = self.get_streamable_weights_size()
+
+        self.engine.weight_streaming_budget_v2 = budget_bytes
+        if self.get_weight_streaming_budget() != budget_bytes:
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py:171]: wouldn't this always be budget_bytes since `self.engine.weight_streaming_budget_v2 = budget_bytes` is being done before this ?

File: tests/py/dynamo/runtime/test_004_weight_streaming.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+            fx_graph,
+            inputs=input,
+            ir="dynamo",
+            min_block_size=1,
+            debug=True,
+            use_python_runtime=use_python_runtime,
+            enable_weight_streaming=True,
+        )
+        # Weight streaming budget is applied manually.
+        ws_context = torchtrt.runtime.weight_streaming_context(optimized_model)
```
Comment (narendasan) [inline tests/py/dynamo/runtime/test_004_weight_streaming.py:96]: Can we use the context manager syntax to use this? 

```py
with torch_tensorrt.runtime.weight_streaming(model) as weight_streaming_ctx:
    current_budget = weight_streaming_ctx.device_budget
    weight_streaming_ctx.device_budget = current_budget * 0.7 # Can add listeners to __setattr__ to trigger functions
    optimized_model(*input)
```

**Review by peri044 — COMMENTED** [review]
Thoughts :
1) If we use weight streaming as default, is there any problem with perf ? assuming we don't allocate any budget or if automatic is chosen, and the model can fit on GPU memory completely

File: core/runtime/TRTEngine.cpp [hunk 12]
```diff
@@ -248,6 +254,43 @@ void TRTEngine::set_profiling_paths() {
   cuda_graph_debug_path = std::filesystem::path{profile_path_prefix + "/" + name + "_cudagraph.dot"}.string();
 }
 
+int64_t TRTEngine::get_weight_streaming_budget_v2() {
+  return cuda_engine->getWeightStreamingBudgetV2();
+}
+
+bool TRTEngine::set_weight_streaming_budget_v2(int64_t budget) {
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp:261]: Lets make these names more clear to people who have never used TRT before. 

Would something like `TRTEngine::set_device_memory_budget` make sense, does that accurately describe the setting here?

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py (hunk 10 above)
Comment (keehyuna) [reply py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py:171]: This [api](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Engine.html#tensorrt.ICudaEngine) Set and get the current weight streaming budget for inference. This [test case](https://github.com/pytorch/TensorRT/pull/3111/files#diff-f9e9b91837ae150afd9f6637efb856ac105a78ec762851d0477c80a3e117dd82R141-R147) verifies it

File: core/runtime/TRTEngine.cpp (hunk 12 above)
Comment (keehyuna) [reply core/runtime/TRTEngine.cpp:261]: I agree with that. There is device_memory_size api in TRT api, it might be confusing.
https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Engine.html

what about to name it to set_weight_streaming_budget()?

File: core/runtime/TRTEngine.cpp [hunk 13]
```diff
@@ -248,6 +254,43 @@ void TRTEngine::set_profiling_paths() {
...
+  return cuda_engine->setWeightStreamingBudgetV2(budget);
+}
+
+// Returns 0 if BuilderFlag::kWEIGHT_STREAMING is unset during engine building.
+int64_t TRTEngine::get_streamable_weights_size() {
```
[the remark this answers is not in the document]
Comment (keehyuna) [reply core/runtime/TRTEngine.cpp:266]: There are v1 and v2 api and I only used v2 api (trt>=10.1). I will share some note to compare them.

File: py/torch_tensorrt/runtime/_weight_streaming.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            for streamable_bytes in self.streamable_budget
+        ]
+        for i, (name, rt_mod) in enumerate(self.rt_mods):
+            ws_budget_bytes += rt_mod.set_device_memory_budget(normalized_size[i])
+            logger.debug(f"Set weight streaming size {normalized_size[i]} for {name}")
+        return ws_budget_bytes
+
+
+def weight_streaming(
+    module: PythonTorchTensorRTModule | TorchTensorRTModule,
```
Comment (narendasan) [inline py/torch_tensorrt/runtime/_weight_streaming.py:72]: This type would likely be a fx.GraphModule instead of just a pure TorchTensorRTModule

**Comment by peri044** [comment]
Model device budget: 
- Include pytorch subgraph weights as well in the device budget (count parameters and estimate size) 
- Conclusion is estimate new budget by subtracting pytorch submodule sizes. If the pytorch weights size is greater than the user provided budget, error out and inform user. 

User API:
1) Enable weight streaming flag 2) device budget

**Comment by narendasan** [comment]
User provides: total_budget for model 

trt_engine_budget = total_budget - sum([p.numel * p.element_size() for p in model.parameters()])

Then we distribute trt_engine_budget proportionally to all engines

Error if trt_engine_budget is <=0 or if any one engine's allocation is less than its min_device_budget

File: py/torch_tensorrt/dynamo/conversion/impl/cat.py [hunk 15]
```diff
@@ -34,6 +34,26 @@ def cat(
...
+                    _enums.dtype._from(promoted_type).to(torch.dtype),
+                    _enums.dtype._from(each_input.dtype).to(torch.dtype),
+                )
+            )
+
+        trt_promoted_type = promoted_type.to(trt.DataType)
+        trt_casted_inputs = []
+        for i, each_input in enumerate(trt_inputs):
+            casted_input = cast_trt_tensor(
+                ctx, each_input, trt_promoted_type, f"{name}_input_casted_{i}"
```
[the remark this answers is not in the document]
Comment (keehyuna) [reply py/torch_tensorrt/dynamo/conversion/impl/cat.py:52]: I thought trt can optimize the perf for relaxed precision. But it seems multiple inputs in ops are eventually casted to same type. Tested sd unet model with/without promoted types, there was no differences. I will generalize.

**Review by keehyuna — COMMENTED** [review]
cpp RT module is in progress.

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 16]
```diff
@@ -88,6 +88,7 @@ def compile(
     engine_cache_dir: Optional[str] = _defaults.ENGINE_CACHE_DIR,
     engine_cache_size: Optional[int] = _defaults.ENGINE_CACHE_SIZE,
     custom_engine_cache: Optional[BaseEngineCache] = _defaults.CUSTOM_ENGINE_CACHE,
+    enable_weight_streaming: bool = _defaults.WEIGHT_STREAMING,
```
Comment (HolyWu) [inline py/torch_tensorrt/dynamo/_compiler.py:91]: The variable name in _defaults should be `ENABLE_WEIGHT_STREAMING`.

File: examples/dynamo/weight_streaming_example.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    llama2_ep,
+    inputs=input_tensors,
+    enabled_precisions={torch.float32},
+    truncate_double=True,
+    device=DEVICE,
+    enable_weight_streaming=True,
+)
+
+# Warm up for 3 iterations
+_ = time_generate(trt_model, input_tensors, osl, 3)
```
Comment (peri044) [inline examples/dynamo/weight_streaming_example.py:102]: How fast did it run for you ? It was slow for me in the past and hence I used only 1 warm up iteration

File: examples/dynamo/weight_streaming_example.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    # Set and get the current weight streaming budget for inference
+    weight_streaming_ctx.device_budget = requested_budget
+    mean_latency = time_generate(trt_model, input_tensors, osl, 1)
+    weight_budget_pct = (
+        weight_streaming_ctx.device_budget
+        / weight_streaming_ctx.total_device_budget
+        * 100
+    )
+    print(
+        f"Set auto weight streaming budget as {weight_budget_pct}%. {weight_streaming_ctx.device_budget} bytes out of {weight_streaming_ctx.total_device_budget}. mean latency = {mean_latency} ms"
```
Comment (peri044) [inline examples/dynamo/weight_streaming_example.py:145]: Would the values here be same as line 117 ? If we enable_weight_streaming=True but don't set automatic budget explicitly ?

File: py/torch_tensorrt/dynamo/conversion/impl/matmul.py [hunk 19]
```diff
@@ -48,6 +49,17 @@ def matrix_multiply(
...
+    promoted_type = _enums.dtype._from(
+        torch.promote_types(
+            _enums.dtype._from(input.dtype).to(torch.dtype),
+            _enums.dtype._from(other.dtype).to(torch.dtype),
+        )
+    )
+    trt_promoted_type = promoted_type.to(trt.DataType)
+    input = cast_trt_tensor(ctx, input, trt_promoted_type, f"{name}_input_casted")
+    other = cast_trt_tensor(ctx, other, trt_promoted_type, f"{name}_other_casted")
+
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/impl/matmul.py:62]: why are these casts needed ? Ideally, they should be already in the correct types? Did you encounter any failure cases ?

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 20]
```diff
@@ -111,6 +111,41 @@ def __init__(
...
+        budget_bytes = self._set_device_memory_budget(budget_bytes)
+        self.context = self.engine.create_execution_context()
+        return budget_bytes
+
+    def _set_device_memory_budget(self, budget_bytes: int) -> int:
+        # Disable weight streaming for invalid budget size
+        if budget_bytes < 0:
+            budget_bytes = self.get_streamable_weights_size()
+        self.engine.weight_streaming_budget_v2 = budget_bytes
+        if self.engine.weight_streaming_budget_v2 != budget_bytes:
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:136]: when would this condition fail ?

File: py/torch_tensorrt/dynamo/conversion/impl/matmul.py (hunk 19 above)
Comment (keehyuna) [reply py/torch_tensorrt/dynamo/conversion/impl/matmul.py:62]: Without this cast, SD UNet model failed with below error. This was caused strongly type network config.
torch_tensorrt [TensorRT Conversion Context][ERROR]ITensor::getDimensions: Error Code 4: API Usage Error ([MATRIX_MULTIPLY]-[torchtrt_lowered_ops.scaled_dot_product_attention]-[scaled_dot_product_attention_1_mm]: IMatrixMultiplyLayer must have same input types. `A` is of type Float and `B` is of type Half.)

I think I can add below condition
`if ctx.net.get_flag(trt.NetworkDefinitionCreationFlag.STRONGLY_TYPED)`

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py (hunk 20 above)
Comment (keehyuna) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:136]: if invalid value(negative, >streamable size) is set, this condition fails.

File: examples/dynamo/weight_streaming_example.py (hunk 17 above)
Comment (keehyuna) [reply examples/dynamo/weight_streaming_example.py:102]: It took 5 seconds in my test with RTX4080.

File: core/runtime/TRTEngine.cpp [hunk 21]
```diff
@@ -260,6 +266,36 @@ void TRTEngine::set_profiling_paths() {
...
+  // Recreating the context because weight streaming budget cannot be modified while there are active context.
+  if (exec_ctx.get() != nullptr) {
+    exec_ctx.reset();
+  }
+  if (profile_execution) {
+    trt_engine_profiler.reset();
+  }
+  bool result = cuda_engine->setWeightStreamingBudgetV2(budget);
+  exec_ctx = make_trt(cuda_engine->createExecutionContext());
+  TORCHTRT_CHECK((exec_ctx.get() != nullptr), "Unable to recreate TensorRT execution context");
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp:283]: Unable to recreate TensorRT execution context __after setting new device memory budget__

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 22]
```diff
@@ -85,6 +85,12 @@ def __init__(
         EXPLICIT_BATCH = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
         flag |= EXPLICIT_BATCH
 
+        if compilation_settings.enable_weight_streaming:
+            STRONGLY_TYPED = 1 << (int)(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:89]: We should log this at least since it affects the graph being created

**Review by narendasan — COMMENTED** [review]
Super minor stuff at this point, think its almost ready to go

**Review by narendasan — APPROVED** [review]
I think this mostly looks good to me, anything outstanding?

**Comment by keehyuna** [comment]
No pending items. I think this PR can be merged.

File: examples/dynamo/weight_streaming_example.py [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+osl = 256
+
+input_tensors = [torch.randint(0, 5, (1, isl), dtype=torch.int64).cuda()]
+model = model.half()
+with torch.no_grad():
+    seq_len = torch.export.Dim("seq_len", min=1, max=osl)
+    # strict=False only enables aotautograd tracing and excludes dynamo.
+    llama2_ep = torch.export.export(
+        model, tuple(input_tensors), dynamic_shapes=({1: seq_len},), strict=False
+    )
```
Comment (peri044) [inline examples/dynamo/weight_streaming_example.py:82]: Can you use the export_llm utility in the utils.py for this ?

**Review by peri044 — CHANGES_REQUESTED** [review]
Minor comments: 
1) Added a comment in the example
2) Also update this example reference in the docsrc/infex.rst to get rendered.
3) Rebase with main to resolve conflicts. 

Overall, changes LGTM

File: examples/dynamo/weight_streaming_example.py (hunk 23 above)
Comment (keehyuna) [reply examples/dynamo/weight_streaming_example.py]: thanks, I updated.
